java-checks/src/main/java/org/sonar/java/checks/ValueBasedObjectsShouldNotBeSerializedCheck.java

```diff
@@ -19,35 +19,35 @@
  */
 package org.sonar.java.checks;
 
+import com.google.common.collect.ImmutableList;
+import java.util.List;
 import org.sonar.check.Rule;
+import org.sonar.java.ast.visitors.SubscriptionVisitor;
 import org.sonar.java.checks.helpers.ValueBasedUtils;
 import org.sonar.java.checks.serialization.SerializableContract;
 import org.sonar.java.model.ModifiersUtils;
+import org.sonar.java.resolve.JavaType;
 import org.sonar.java.resolve.ParametrizedTypeJavaType;
-import org.sonar.plugins.java.api.JavaFileScanner;
-import org.sonar.plugins.java.api.JavaFileScannerContext;
 import org.sonar.plugins.java.api.semantic.Type;
-import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
 import org.sonar.plugins.java.api.tree.ClassTree;
 import org.sonar.plugins.java.api.tree.Modifier;
 import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.Tree.Kind;
 import org.sonar.plugins.java.api.tree.VariableTree;
 
 @Rule(key = "S3437")
-public class ValueBasedObjectsShouldNotBeSerializedCheck extends BaseTreeVisitor implements JavaFileScanner {
+public class ValueBasedObjectsShouldNotBeSerializedCheck extends SubscriptionVisitor {
   
-  private JavaFileScannerContext context;
-
   @Override
-  public void scanFile(JavaFileScannerContext context) {
-    this.context = context;
-    scan(context.getTree());
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.CLASS);
```

Comment:
ENUM INTERFACE: yes, this check applies. An enum is automatically serializable, so its fields, if any, should be checked against value-based nature.
ANNOTATION_TYPE: yes, this check applies, although serializing a serializable annotation instance is an extreme case!